libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp

```diff
@@ -0,0 +1,469 @@
+
+            prev_state = state;
+            state = std::allocate_shared<detail::async_rw_mutex_shared_state,
+                Allocator>(alloc);
+            prev_access = detail::async_rw_mutex_access_type::readwrite;
+            return {std::move(prev_state), state};
+        }
+
+    private:
+        template <detail::async_rw_mutex_access_type AccessType>
+        struct sender
+        {
+            std::shared_ptr<detail::async_rw_mutex_shared_state> prev_state;
+            std::shared_ptr<detail::async_rw_mutex_shared_state> state;
+
+            template <template <typename...> class Tuple,
+                template <typename...> class Variant>
+            using value_types =
+                Variant<Tuple<detail::async_rw_mutex_access_wrapper<
+                    readwrite_type, read_type, AccessType>>>;
+
+            template <template <typename...> class Variant>
+            using error_types = Variant<std::exception_ptr>;
+
+            static constexpr bool sends_done = false;
+
+            template <typename R>
+            struct operation_state
+            {
+                std::decay_t<R> r;
+                std::shared_ptr<detail::async_rw_mutex_shared_state> prev_state;
+                std::shared_ptr<detail::async_rw_mutex_shared_state> state;
+
+                operation_state(operation_state&&) = delete;
+                operation_state(operation_state const&) = delete;
+
+                void start() noexcept
+                {
+                    HPX_ASSERT_MSG(prev_state,
+                        "async_rw_lock::sender::operation_state prev_state is "
+                        "empty, was the sender already started?");
+                    HPX_ASSERT_MSG(state,
+                        "async_rw_lock::sender::operation_state state is "
+                        "empty, was the sender already started?");
+
+                    prev_state->add_continuation(
+                        [r = std::move(r), state = std::move(state)]() mutable {
+                            try
+                            {
+                                hpx::execution::experimental::set_value(
+                                    std::move(r),
+                                    detail::async_rw_mutex_access_wrapper<
+                                        readwrite_type, read_type, AccessType>{
+                                        std::move(state)});
+                            }
+                            catch (...)
+                            {
+                                hpx::execution::experimental::set_error(
+                                    std::move(r), std::current_exception());
+                            }
+                        });
+
+                    // We release prev_state here to allow continuations to run.
+                    // The operation state may otherwise keep it alive longer
+                    // than needed.
+                    prev_state.reset();
```

Comment:
Concurrent invocations of `start` may lead to unreported inconsistencies such that the `HPX_ASSERT` does not fire even if another thread is trying to execute `start` at the same time. I'm not sure if this is a real problem, though.

Reply:
Can you explain how? `start` can only be called once on a (connected) sender.